data-prepper-plugins/common/src/main/java/org/opensearch/dataprepper/common/sink/DefaultSinkOutputStrategy.java

```diff
@@ -0,0 +1,107 @@
+import static org.opensearch.dataprepper.logging.DataPrepperMarkers.NOISY;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.List;
+
+public abstract class DefaultSinkOutputStrategy {
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultSinkOutputStrategy.class);
+    private final LockStrategy lockStrategy;
+    private final SinkBuffer sinkBuffer;
+    private final SinkMetrics sinkMetrics;
+    private final SinkFlushContext sinkFlushContext;
+
+    public DefaultSinkOutputStrategy(final LockStrategy lockStrategy, final SinkBuffer sinkBuffer, final SinkFlushContext sinkFlushContext, final SinkMetrics sinkMetrics) {
+        this.lockStrategy = lockStrategy;
+        this.sinkBuffer = sinkBuffer;
+        this.sinkMetrics = sinkMetrics;
+        this.sinkFlushContext = sinkFlushContext;
+    }
+
+    public void flushBuffer() {
+        long startTime = System.nanoTime();
+        // getBuffer() should return the buffer contents
+        SinkFlushableBuffer flushableBuffer = sinkBuffer.getFlushableBuffer(sinkFlushContext);
+        try {      
+            SinkFlushResult flushResult = flushableBuffer.flush();
+            if (flushResult == null) { // success
+                sinkMetrics.recordRequestLatency((double)(System.nanoTime() - startTime));
+                List<Event> events = flushableBuffer.getEvents();
+                for (final Event event: events) {
+                    event.getEventHandle().release(true);
+                }
+            } else {    
+                // flush Result should contain the events that are 
+                // failed to be delivered, so that these events can be forwarded to DLQ
+                addFailedEventsToDLQ(flushResult.getEvents(), flushResult.getException());
+            }           
+        } catch (Exception e) {
+            // Add list of events to DLQ
+            addFailedEventsToDLQ(flushableBuffer.getEvents(), e);
+        }              
+    }
+
+    public void execute(Collection<Record<Event>> records) {
+        lockStrategy.lock();
+        try {
+            if (sinkBuffer.exceedsFlushTimeInterval()) {
+                flushBuffer();
+            }
+            if (records == null || records.isEmpty()) {
+                return;
+            }
+            for (Record<Event> record : records) {
+                final Event event = record.getData();
+                try {
+                    // getSinkBufferEntry() is a sink method that may use codec to get
+                    // the estimated size of the event
+                    SinkBufferEntry bufferEntry = getSinkBufferEntry(event);
+
+                    // Check if individual event exceeds sink's max event size
+                    if (exceedsMaxEventSizeThreshold(bufferEntry.getEstimatedSize())) {
+                        throw new RuntimeException("Event size exceeds max allowed event size");
```

Comment:
Wouldn't this lead to a dropped event that needs counting?

Reply:
Not necessarily. It will be sent to DLQ. If DLQ handling code sees DLQ is not configured, then it will be dropped and counter should be incremented.